contrib/codegen/src/database.rs

```diff
@@ -0,0 +1,133 @@
+const ONLY_ON_STRUCTS_MSG: &str = "`database` attribute can only be used on structs";
+const ONLY_UNNAMED_FIELDS: &str = "`database` attribute can only be applied to structs with \
+    exactly one unnamed field";
+const NO_GENERIC_STRUCTS: &str = "`database` attribute cannot be applied to a struct with a \
+    generic type";
+
+fn parse_invocation(attr: TokenStream, input: TokenStream) -> Result<DatabaseInvocation> {
+    let attr_stream2 = ::proc_macro2::TokenStream::from(attr);
+    let attr_span = attr_stream2.span();
+    let string_lit = ::syn::parse2::<LitStr>(attr_stream2)
+        .map_err(|_| attr_span.error("expected string literal"))?;
+
+    let input = ::syn::parse::<DeriveInput>(input).unwrap();
+    if !input.generics.params.is_empty() {
+        return Err(input.span().error(NO_GENERIC_STRUCTS));
+    }
+
+    let structure = match input.data {
+        Data::Struct(s) => s,
+        _ => return Err(input.span().error(ONLY_ON_STRUCTS_MSG))
+    };
+
+    let inner_type = match structure.fields {
+        Fields::Unnamed(ref fields) if fields.unnamed.len() == 1 => {
+            let first = fields.unnamed.first().expect("checked length");
+            first.value().ty.clone()
+        }
+        _ => return Err(structure.fields.span().error(ONLY_UNNAMED_FIELDS).help(EXAMPLE))
+    };
+
+    Ok(DatabaseInvocation {
+        type_name: input.ident,
+        db_name: string_lit.value(),
+        structure: structure,
+        connection_type: inner_type,
+    })
+}
+
+pub fn database_attr(attr: TokenStream, input: TokenStream) -> Result<TokenStream> {
+    let invocation = parse_invocation(attr, input)?;
+
+    let connection_type = &invocation.connection_type;
+    let database_name = &invocation.db_name;
+    let request_guard_type = &invocation.type_name;
+    let pool_type = Ident::new(&format!("{}Pool", request_guard_type), request_guard_type.span());
+
+    // The Rocket config has to be cloned because otherwise without `#![feature(nll)]`, you end up
+    // with an error about not being able to move Rocket when adding to managed state or returning
+    // an `Err(rocket)` since it's borrowed by the `rocket.config()` method call.
+    let tokens = quote! {
+        pub struct #request_guard_type(
+            pub ::rocket_contrib::databases::r2d2::PooledConnection<<#connection_type as ::rocket_contrib::databases::Poolable>::Manager>
+        );
+        pub struct #pool_type(::rocket_contrib::databases::r2d2::Pool<<#connection_type as ::rocket_contrib::databases::Poolable>::Manager>);
+
+        impl #request_guard_type {
+            pub fn fairing() -> impl ::rocket::fairing::Fairing {
+                use ::rocket_contrib::databases::Poolable;
+
+                ::rocket::fairing::AdHoc::on_attach(|rocket| {
+                    let pool = ::rocket_contrib::databases::database_config(#database_name, &rocket.config())
+                        .map(#connection_type::pool);
+
+                    match pool {
```

Comment:
This still feels cumbersome and verbose for how I match out the errors. I felt this was the simplest way. Handling the error piecemeal, as you recommended, resulted in borrowck errors, so I kept with the original approach but made sure to print the error out.

I also make the assumption that the user has `#[macro_use]` for `rocket` and have also included the `log` crate. This may be too big of an assumption to make, though. I'm undecided on if that is the right way to go about this.

Reply:
> I also make the assumption that the user has #[macro_use] for rocket and have also included the log crate.

We can't make this assumption. This is why we wrote the `pretty_print` method.